### PR TITLE
refactor(providers): separate resolving links from scraping method

### PR DESCRIPTION
**Pull request by ddivad**
## refactor(tv): separate resolving links from scraping method

Update TV providers to separate the resolving links block of code from the `scrape` method in each provider.

The reason for this separation is to aid with resolving cached links. When links are cached, the url to the video (before resolving) will be stored. This means that when cached links are to be resolved, an array of links will be returned from the cache, and the `scrape` function of each provider can be skipped.

Separating out the resolve links piece of the providers, and letting it accept an array of urls as a parameter will make this piece of code easier to reuse between the case where cached links are available and when the providers have to be scraped to find links.

### What's in this PR:
**BaseProvider:**
- Add new function `resolveVideoLinks` to `BaseProvider` that should be implemented in each subclass.
- Rename `resolveLinks` to `_resolveLinks`.
- Add `remoteAddress` to `BaseProvider`.

**TV Providers:**
- Update TV providers to use the changes implemented in #42.
- Implement `resolveVideoLinks` in each TV provider (movie providers will follow).

### CODE REVIEW

**Comment by ddivad** [comment]
@SpajicM @Teekzilla First 'caching related' PR if you are interested in taking a look. It adds the last pieces of foundations (for tv providers anyway).

File: src/scrapers/providers/tv/GoWatchSeries.js [hunk 1]
```diff
@@ -60,34 +54,37 @@ module.exports = class SwatchSeries extends BaseProvider {
...
-            iframeLinks.forEach((link) => {
-                const headers = {
-                    'user-agent': userAgent,
-                    'x-real-ip': clientIp,
-                    'x-forwarded-for': clientIp
-                };
-                resolvePromises.push(this.resolveLink(link, ws, jar, headers));
-            });
-            
+            resolvePromises = this.scrapeVideoLinks(ws, videoUrls)
```
Comment (SpajicM) [inline src/scrapers/providers/tv/GoWatchSeries.js]: resolveVideoLinks*

**Comment by SpajicM** [comment]
resolveVideoLinks method seems the same for all providers, so should it really be implemented for each provider or it could be shared in BaseProvider. I think only thing that could vary by provider could be headers.

**Comment by Teekzilla** [comment]
Look good to me.

> resolveVideoLinks method seems the same for all providers, so should it really be implemented for each provider or it could be shared in BaseProvider. I think only thing that could vary by provider could be headers.

Agreed here,  I don't think there is necessarily a need to have `resolveVideoLinks` implemented in each provider. Structure is the same, and I actually don't even know if the headers would change that much if you already have retrieved the embed links would they?

**Comment by ddivad** [comment]
@SpajicM @Teekzilla Thanks for the suggestions.

Regarding the `resolveVideoLinks` method being the same everywhere - this is widely true, but there are some specific cases where it is slightly different. Eg: `SeriesFree`

For SeriesFree, we need to call the `scrapeHarder` function for each link, and this makes the `resolveLinks` function different in this case. 

I think what I'll do though is add it to the BaseProvider, and then in specific cases like that I can just override in the specific provider that needs a more specific implementation. I guess that covers both scenarios pretty well then.

**Pull request closed without merging**